File: MAXED/fillfil.pl
#!/usr/bin/perl

sub FILLFIL{
	local *ENBZKL=shift;
	local *ZKL=shift;
	local *N=shift;
	local *N0=shift;
	local *NB=shift;
	local *ENBF=shift;
	local *FI=shift;
	
	my $K,$L,$N0M1,$R1,$R2;
	my @FIL,@ZKLL;

	$N0M1 = $N0 - 1;

#print " FILLFIL stop 1\n";
	for ($K=1;$K<$N+1;$K++){
		$FIL[$K] = 0.0;
	}

#print " FILLFIL stop 2\n";
	for ($K=1;$K<$N0M1+1;$K++){
		$ZKLL[$K] = $ZKL[$K]/(log($ENBZKL[$K+1]) - log($ENBZKL[$K]));
	}  

#print " FILLFIL stop 3\n";
	for ($K=1;$K<$N0M1+1;$K++){
#print " FILLFIL stop 4\n";
		for ($L=1;$L<$NB+1;$L++){
	  		$R2 = log($ENBF[$L+1]) - log($ENBF[$L]); 
	  		if ( ($ENBF[$L]) <= ($ENBZKL[$K]) ) {
	    		if ( ($ENBF[$L+1]) > ($ENBZKL[$K]) ) { 
	      			if ( ($ENBF[$L+1])  >= ($ENBZKL[$K+1]) ) {
						$R1 = log($ENBZKL[$K+1]) - log($ENBZKL[$K]);
						$FIL[$L] = $FIL[$L] + $ZKLL[$K]*($R1/$R2);
					}
	      			elsif ( ($ENBF[$L+1]) < ($ENBZKL[$K+1]) ) {
						$R1 = log($ENBF[$L+1]) - log($ENBZKL[$K]) ;
						$FIL[$L] = $FIL[$L] + $ZKLL[$K]*($R1/$R2);
	      			}
	   			}
	  		}
			elsif ( ($ENBF[$L]) > ($ENBZKL[$K]) ){
	    		if ( ($ENBF[$L]) < ($ENBZKL[$K+1]) ){ 
	      			if ( ($ENBF[$L+1]) <= ($ENBZKL[$K+1])) {
						$R1 = log($ENBF[$L+1]) - log($ENBF[$L]);
						$FIL[$L] = $FIL[$L] + $ZKLL[$K]*($R1/$R2);
					}
	      			elsif ( ($ENBF[$L+1]) > ($ENBZKL[$K+1])){
						$R1 = log($ENBZKL[$K+1]) - log($ENBF[$L]);
						$FIL[$L] = $FIL[$L] + $ZKLL[$K]*($R1/$R2);
					}
				}
			}
		}
	}
#print " FILLFIL stop 5\n";

	for ($K=1;$K<$NB+1;$K++){
		$FI[$K] = $FIL[$K]*(log($ENBF[$K+1]) - log($ENBF[$K]));
	}  

	return;
}
1;

File: bums2/FORTRAN_METHODS/MAXED/bins/fillfil.py
#This file will be the equivalent of fillfil.pl from the original MAXED directory
import numpy as np

class SpectrumBinFiller:
    def __init__(self, enbzkl, zkl, enbf):
        self.enbzkl = np.asarray(enbzkl, dtype=np.float64)
        self.zkl = np.asarray(zkl, dtype=np.float64)
        self.enbf = np.asarray(enbf, dtype=np.float64)

        self.n0 = len(enbzkl)
        self.nb = len(enbf) - 1
        self.n = len(enbf) - 1

        self.fil = np.zeros(self.nb, dtype=np.float64)
        self.fi = np.zeros(self.nb, dtype=np.float64)

    def fill(self):
        n0m1 = self.n0 - 1

        #Precompute ZKLL, the log scaled bin height
        zkll = self.zkl[:n0m1] / (np.log(self.enbzkl[1:self.n0]) - np.log(self.enbzkl[:n0m1]))

        #Main loop from the original Perl
        for k in range(n0m1):
            for l in range(self.nb):
                r2 = np.log(self.enbf[l + 1]) - np.log(self.enbf[l])

                #Case 1: ENBF[l] ≤ ENBZKL[k] < ENBF[l+1]
                if self.enbf[l] <= self.enbzkl[k] < self.enbf[l + 1]:
                    if self.enbf[l + 1] >= self.enbzkl[k + 1]:
                        r1 = np.log(self.enbzkl[k + 1]) - np.log(self.enbzkl[k])
                        self.fil[l] += zkll[k] * (r1/r2)
                    elif self.enbf[l + 1] < self.enbzkl[k + 1]:
                        r1 = np.log(self.enbf[l + 1]) - np.log(self.enbzkl[k])
                        self.fil[l] += zkll[k] * (r1/r2)
                
                #Case 2: ENBF[l] > ENBZKL[k]
                elif self.enbf[l] > self.enbzkl[k]:
                    if self.enbf[l] < self.enbzkl[k + 1]:
                        if self.enbf[l + 1] <= self.enbzkl[k + 1]:
                            r1 = np.log(self.enbf[l + 1]) - np.log(self.enbf[l])
                            self.fil[l] += zkll[k] * (r1/r2)
                        elif self.enbf[l + 1] > self.enbzkl[k + 1]:
                            r1 = np.log(self.enbzkl[k + 1]) - np.log(self.enbf[l])
                            self.fil[l] += zkll[k] * (r1/r2)

        #Final adjustment to get FI
        for h in range(self.nb):
            width = np.log(self.enbf[h + 1]) - np.log(self.enbf[h])
            self.fi[h] = self.fil[h] * width

            
        return self.fi

        